### PR TITLE
Fix and improve the efficiency of the backward of xlog* functions.

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82714
* __->__ #82713

That is `xlogy`, `special.xlogy`, `special.xlog1py`.

Fixes https://github.com/pytorch/pytorch/issues/80770
Fixes https://github.com/pytorch/pytorch/issues/74279

### CODE REVIEW

File: test/distributions/test_distributions.py [hunk 1]
```diff
@@ -1424,9 +1424,6 @@ def ref_log_prob(ref_rate, idx, x, log_prob):
         # approximation enters the forbidden parameter space. We instead compare with the
         # theoretical results.
         dist = Poisson(rate_zero)
-        s = dist.sample()
-        dist.log_prob(s).backward()
-        torch.testing.assert_allclose(rate_zero.grad, -1.0)
```
Comment (Lezcano) [inline test/distributions/test_distributions.py:1429]: This test was wrong. The pdf is not differentiable at zero.

File: tools/autograd/derivatives.yaml [hunk 2]
```diff
@@ -911,29 +911,29 @@
   result: self_t / (1 + pow(2, other_p - self_p)) + other_t / (1 + pow(2, self_p - other_p))
 
 - name: xlogy.Tensor(Tensor self, Tensor other) -> Tensor
-  self: grad * at::xlogy((self != 0), other)
-  other: grad * at::where(other.isnan() | (self != 0), self / other, zeros_like(other))
-  result: self_t * at::xlogy((self_p != 0), other_p) + other_t * self_p / other_p
+  self: at::xlogy(grad, other)
+  other: grad * self / other
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:915]: This changes the value when other is nan or zero? Is that ok?

Comment (Lezcano) [reply tools/autograd/derivatives.yaml:915]: When it's NaN we can return whatever we feel like. The previous value when `self` is zero was wrong and this fixes it.

Comment (albanD) [reply tools/autograd/derivatives.yaml:915]: > When it's NaN we can return whatever we feel like.

That's not entirely true. We do try to preserve NaNs usually :p

**Review by albanD — COMMENTED** [review]
Can you add one-off tests for the gradients at the non-differentiable points to make sure we don't change these in the future by mistake?

File: tools/autograd/derivatives.yaml (hunk 2 above)
Comment (Lezcano) [reply tools/autograd/derivatives.yaml:915]: Oh, in that case this implementation is better than the previous one, which put zeros in the place of nans lol

Comment (ngimel) [reply tools/autograd/derivatives.yaml:915]: @alband in backward we don't (if forward function is not defined), we've had an issue about log recently, when someone was complaining of getting non-nan gradients for negative inputs.

**Comment by Lezcano** [comment]
@albanD added the test. When we decide on how to go with https://github.com/pytorch/pytorch/issues/80770#issuecomment-1216481516, this PR should be ready to go.

**Comment by Lezcano** [comment]
@albanD I implemented the discussion in https://github.com/pytorch/pytorch/issues/80770#issuecomment-1216769221 and left a comment. PTAL.

**Review by albanD — APPROVED** [review]
Perfect!
Thanks for the thorough update

**Pull request closed without merging**